### PR TITLE
Adding logging capabilities to the curl requests

### PR DESCRIPTION
**Pull request by james-rowlands**
Adding logging capabilities to the curl requests so we have visibility of requests, responses, http codes, and the duration of a request.

### CODE REVIEW

**Comment by wyliedRentalcars** [comment]
great idea lets the braintree php library support custom logging code on the client side.  Should let the library be auditable.

**Comment by wjr1985** [comment]
Thanks for the contribution @james-rowlands! We think adding logging to the cURL requests is a great idea and will be reviewing this PR in the near future.

**Comment by wjr1985** [comment]
Would it be possible for you to include a default logger that implements your defined interface, but can be used outside of testing? This would more closely reflect how logging is implemented in other client libraries. If you're looking for inspiration for the output format, it could match what our Ruby CL outputs:

`[Braintree] [<DATE_OF_REQUEST>] [<REQUEST METHOD>] [<PATH>] [<REQUEST RESPONSE CODE>]`

Here's an example:

`[Braintree] [09/Dec/2015 17:41:59 UTC] GET /merchants/a1b2c3d4e5f6/customers/12345678 200`

I'll keep reviewing the PR and let you know any other comments on the files themselves.

File: lib/Braintree/Configuration.php [hunk 1]
```diff
@@ -267,6 +288,25 @@ public function setPrivateKey($value)
...
+    /**
+     * @return CurlAuditLoggerInterface
+     */
+    public function getLogger()
+    {
+        return $this->_logger;
+    }
+
+    /**
+     * Do not use this method directly. Pass in the merchantId to the constructor.
```
Comment (wjr1985) [inline lib/Braintree/Configuration.php]: Could you change `merchantId` to `logger` in this comment?

File: lib/Braintree/Logging/CurlAuditLoggerInterface.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
+namespace Braintree\Logging;
+
+/**
+ * Raised when authentication fails.
+ * This may be caused by an incorrect Configuration
```
Comment (wjr1985) [inline lib/Braintree/Logging/CurlAuditLoggerInterface.php]: Could you change these two comment lines to reflect what the interface is for?

File: lib/Braintree/Logging/CurlAuditLoggerInterface.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * @subpackage Logging
+ */
+interface CurlAuditLoggerInterface
+{
+
+
+    public function log();
+
+    /**
+     * @param $responseTime
```
Comment (wjr1985) [inline lib/Braintree/Logging/CurlAuditLoggerInterface.php]: For all the methods defined in this file, could you document what they are expected to do, and add some additional documentation for each variable?

**Comment by matthewpatterson** [comment]
Hey @james-rowlands,

I really like this idea, but I think there's a way to make it even better. I think it would make sense to be able to inject any [`Psr\Log\LoggerInterface`](https://github.com/php-fig/log/blob/master/Psr/Log/LoggerInterface.php) into the library. [PSR-3](http://www.php-fig.org/psr/psr-3/) is becoming more widely adopted and it'd be great to allow developers to inject their framework's PSR-3 compatible logger into the Braintree PHP library.

What do you think?

-mattpatt

**Comment by james-rowlands** [comment]
I have made the changes recommended in your comments with one exception, the logging we have implemented is designed to work by inserting the details of the curl request and response into a searchable auditable DB, so our aim is to create a row in a database with an id, request, response and http code 

While a PSR-3 compliant logging implementation would not easily fit into this requirement. And could lead to you using it to log other things in Braintree that are not directly related to the curl request.

One option would be to make the logging variables curlInformationLogger variables so that a PSR-3 compatible logger could be implemented throughout Braintree_php and if possible keep the curl DB logger separate to the PSR-3 compatible logger.

It is possible (and we do) use a PSR-3 compatible logger to be passed into braintree from our end.

The other option would be for you guys to implement Symphony compatible events which we could just hook into from our end.

File: lib/Braintree/Configuration.php [hunk 4]
```diff
@@ -111,6 +116,22 @@ public static function privateKey($value=null)
     }
 
     /**
+     * Sets or gets the logger to use
+     *
+     * @param string $value If provided, sets the logger
```
Comment (dereuromark) [inline lib/Braintree/Configuration.php]: string|null

File: lib/Braintree/Configuration.php [hunk 5]
```diff
@@ -111,6 +116,22 @@ public static function privateKey($value=null)
...
+     * @return string The logger to be used
+     */
+    public static function logger($value=null)
+    {
+        if (empty($value)) {
```
Comment (dereuromark) [inline lib/Braintree/Configuration.php]: Should be `=== null`

File: lib/Braintree/Http.php [hunk 6]
```diff
@@ -144,8 +153,20 @@ public function _doUrlRequest($httpVerb, $url, $requestBody = null)
         }
 
         curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
+
+        if(isset($this->_logger)) {
```
Comment (dereuromark) [inline lib/Braintree/Http.php]: `=== null`

File: lib/Braintree/Http.php [hunk 7]
```diff
@@ -144,8 +153,20 @@ public function _doUrlRequest($httpVerb, $url, $requestBody = null)
...
+            $this->_logger->setRequest($requestBody, $url);
+            $this->_logger->setRequestTime(time());
+        }
+
         $response = curl_exec($curl);
         $httpStatus = curl_getinfo($curl, CURLINFO_HTTP_CODE);
+
+        if(isset($this->_logger)) {
```
Comment (dereuromark) [inline lib/Braintree/Http.php]: cs issue: space missing after if.

**Comment by EvanHahn** [comment]
@james-rowlands Do you want to address the comments and merge `master` into this branch? If not, we will close this pull request.

**Comment by EvanHahn** [comment]
We're going to close this PR for now due to lack of activity. Feel free to reopen a new one!

**Pull request closed without merging**